contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,209 @@
+<frontmatter>
+  title: "Javascript: Modules"
+  footer: footer.md
+  head: head.md
+  siteNav: mainNav.md
+  pageNav: 3
+</frontmatter>
+
+{{ navbar | safe }}
+
+<div class="website-content">
+
+# Javascript: Modules
+
+**Author: Gilbert Emerson**<br>
+Reviewers: Chelsey Ong, Ong Shu Peng, Amrut Prabhu
+
+<box type="info">
+
+This article assumes the reader has some basic knowledge of JavaScript.
+</box>
+
+## What Is a Module?
+
+You might often encounter the term such as _module, package, library, dependency, plugin,_ etc. when developing your own application in any programming languages, including JavaScript. We will not dive into the exact definition of each term because that is not the scope of this article, but we will define a module as a small part of code that is broken up from a larger code base.
+
+Let's say you have an application with functionalities A and B, where functionality A needs functionality B. Without modules, both of these functionalities are mixed together in the code base without a clear separation. With modules, we can separate those 2 functionalities into separate module each. When A need B, A will "include" B and A will be able to work as if A and B has never been separated. The exact usage of modules will be illustrated in the **How to Modularize JavaScript Code** section after the following section.
+
+## Why Use Modules in JavaScript?
+
+Following are reasons why use modules in JavaScript. The reasons below are not exclusive to JavaScript, but may also be general reasons for using modules in other languages.
+
+**1. Better management of code base** <br>
+Using modules will break the code base into smaller parts which if done well can help in managing a code base, especially a large one. To be able to modularize your code well, you can apply Software Engineering principles such as _[SOLID](http://se-education.org/se-book/principles/solidPrinciples/index.html)_, etc.
+
+**2. Namespacing** <br>
+Because of how [JavaScript scoping](https://www.sitepoint.com/demystifying-javascript-variable-scope-hoisting/) works, avoiding global variable can be an issue. Modules allows JavaScript developer to avoid this problem by isolating the namespace to a specific module.
+
+**3. Reusability** <br>
+As per _[Don't Repeat Yourself (DRY) Principle](http://se-education.org/se-book/principles/dryPrinciple/index.html)_, modules allows developers to reuse their code that is contained in a module. If for example we have an application that rely on a fuctionality such as string comparison function, we can separate that function into a module and let the application to use the function from that module instead of having to always repeating that function in all the places where it is needed.
+
```

Comment:
I would like this part to tighten up a bit more. See if the following works:

----

## What Is a Module?

In programming, the term _module_ (other similar terms: _package, library, dependency, plugin,_ etc.) is used to refer to a _small part of code that is broken up from a larger code base_.

Modules help programmers in many ways. Some examples:
* **Modules make the code more manageable**<br>
  Using modules will break the code base into smaller parts which if done well can help in managing a code base, especially a large one. For example, let's say you have an application with functionalities A and B, where functionality A needs functionality B. Without modules, both of these functionalities are mixed together in the code base without a clear separation. With modules, we can separate those 2 functionalities into separate module each. When A need B, A will 'include' B and A will be able to work as if A and B has never been separated. 
* **Modules promote reuse**<br>
  Modules allows developers to reuse their code that is contained in a module. If for example we have an application that rely on a fuctionality such as string comparison function, we can separate that function into a module and let the application to use the function from that module instead of having to always repeating that function in all the places where it is needed.
* **Modules help minimize name clashes**<br>
  Breaking the code into modules result in breaking the code's namespace into smaller parts too; this helps to minimize name clashes and the need for global variables.